src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs

```diff
@@ -387,24 +406,69 @@ internal static SyntaxTree CreateWithoutClone(CSharpSyntaxNode root)
             CSharpParseOptions? options = null,
             string path = "",
             Encoding? encoding = null,
-            ImmutableDictionary<string, ReportDiagnostic>? diagnosticOptions = null,
-            bool? isGeneratedCode = null,
             CancellationToken cancellationToken = default)
         {
+#pragma warning disable CS0618 // We are calling into the obsolete member as that's the one that still does the real work
+            return ParseText(text, options, path, encoding, diagnosticOptions: null, cancellationToken);
+#pragma warning restore CS0618
+        }
+
+#pragma warning restore RS0027
+
+        /// <summary>
+        /// Produces a syntax tree by parsing the source text.
+        /// </summary>
+        /// <param name="diagnosticOptions">An obsolete parameter. Diagnostic options should now be passed with <see cref="CompilationOptions.SyntaxTreeOptionsProvider"/></param>
+        /// <param name="isGeneratedCode">An obsolete parameter. It is unused.</param>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("The diagnosticOptions and isGeneratedCode parameters are obsolete due to performance problems, if you are using them use CompilationOptions.SyntaxTreeOptionsProvider instead", error: true)]
+        public static SyntaxTree ParseText(
+            string text,
+            CSharpParseOptions? options,
+            string path,
+            Encoding? encoding,
+            ImmutableDictionary<string, ReportDiagnostic>? diagnosticOptions,
+            bool? isGeneratedCode,
+            CancellationToken cancellationToken)
+        {
             return ParseText(SourceText.From(text, encoding), options, path, diagnosticOptions, isGeneratedCode, cancellationToken);
         }
 
+        // The overload that has more parameters is itself obsolete, as an intentional break to allow future
+        // expansion
+#pragma warning disable RS0027 // Public API with optional parameter(s) should have the most parameters amongst its public overloads.
+
         /// <summary>
         /// Produces a syntax tree by parsing the source text.
         /// </summary>
         public static SyntaxTree ParseText(
```

Comment:
This API should still be deprecated (not obsolete, but deprecated). We have an overload that takes one more optional parameter (an encoding) that is not obsolete.

Reply:
On enlightenment from Jason, I realize that these have different parameter types. Never mind.

---
In reply to: [460289471](https://github.com/dotnet/roslyn/pull/44331#discussion_r460289471) [](ancestors = 460289471)